libr/lang/p/qjs/io.c

```diff
@@ -0,0 +1,258 @@
+
+	return NULL;
+}
+
+// TODO maybe add a function to call by plugin name? (is 1 extra arg)
+static char *qjs_io_system(RIO *io, RIODesc *fd, const char *input) {
+	RCore *core = io->coreb.core;
+	QjsPluginManager *pm = R_UNWRAP4 (core, lang, session, plugin_data);
+
+	// Iterate over plugins until one returns "true" (meaning the plugin handled the input)
+	QjsIoPlugin *plugin;
+	R_VEC_FOREACH (&pm->io_plugins, plugin) {
+		JSContext *ctx = plugin->ctx;
+		JSValueConst args[1] = { JS_NewString (ctx, input) };
+		JSValue res = JS_Call (ctx, plugin->fn_system_js, JS_UNDEFINED, countof (args), args);
+		if (JS_ToBool (ctx, res)) {
+			return strdup ("true");
+		}
+	}
+
+	return NULL;
+}
+
+static JSValue r2plugin_io(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
+	JSRuntime *rt = JS_GetRuntime (ctx);
+	QjsPluginManager *pm = JS_GetRuntimeOpaque (rt);
+
+	if (argc != 2) {
+		return JS_ThrowRangeError (ctx, "r2.plugin expects two arguments");
+	}
+
+	JSValueConst args[1] = { JS_NewString (ctx, ""), };
+
+	// check if res is an object
+	JSValue res = JS_Call (ctx, argv[1], JS_UNDEFINED, countof (args), args);
+	if (!JS_IsObject (res)) {
+		return JS_ThrowRangeError (ctx, "r2.plugin function must return an object");
+	}
+
+	JSValue name = JS_GetPropertyStr (ctx, res, "name");
+	size_t namelen;
+	const char *nameptr = JS_ToCStringLen2 (ctx, &namelen, name, false);
+	if (!nameptr) {
+		R_LOG_WARN ("r2.plugin requires the function to return an object with the `name` field");
+		return JS_NewBool (ctx, false);
+	}
+
+	JSValue fn_check_js = JS_GetPropertyStr (ctx, res, "check");
+	JSValue fn_open_js = JS_GetPropertyStr (ctx, res, "open");
+	JSValue fn_read_js = JS_GetPropertyStr (ctx, res, "read");
+	JSValue fn_seek_js = JS_GetPropertyStr (ctx, res, "seek");
+
+	if (!JS_IsFunction (ctx, fn_check_js)
+	 || !JS_IsFunction (ctx, fn_open_js)
+	 || !JS_IsFunction (ctx, fn_read_js)
+	 || !JS_IsFunction (ctx, fn_seek_js)
+	 ) {
+		R_LOG_WARN ("r2.plugin('io', X) expects X to be a function that returns an object with at least: check, open, read, seek");
+		// return JS_ThrowRangeError (ctx, "r2.plugin requires the function to return an object with the `call` field to be a function");
+		return JS_NewBool (ctx, false);
+	}
+
+	QjsIoPlugin *cp = plugin_manager_find_io_plugin (pm, nameptr);
+	if (cp) {
+		R_LOG_WARN ("r2.plugin with name %s is already registered", nameptr);
+		// return JS_ThrowRangeError (ctx, "r2.plugin core already registered (only one exists)");
+		return JS_NewBool (ctx, false);
+	}
+
+	RIOPlugin *ap = R_NEW0 (RIOPlugin);
+	if (!ap) {
+		return JS_ThrowRangeError (ctx, "could not allocate qjs core plugin");
+	}
+
+	JSValue desc = JS_GetPropertyStr (ctx, res, "desc");
+	JSValue license = JS_GetPropertyStr (ctx, res, "license");
+	const char *descptr = JS_ToCStringLen2 (ctx, &namelen, desc, false);
+	const char *licenseptr = JS_ToCStringLen2 (ctx, &namelen, license, false);
+
+	ap->name = nameptr;  // TODO no strdup here?
```

Comment:
Depends on if it nameptr is managed by qjs, but imo I would do it, and then just free it in the fini function of QjsIoPlugin